torch/__init__.py

```diff
@@ -426,11 +426,41 @@ def use_deterministic_algorithms(d):
     Note that deterministic operations tend to have worse performance than
     non-deterministic operations.
 
+    .. note::
+
+        :func:`use_deterministic_algorithms` is not guaranteed to prevent
+        nondeterministic behavior that is caused by an operation performing
+        multiple writes of different data to a single memory location. The
+        order of writes to a single location is not guaranteed, which can cause
+        nondeterminism. This could occur when an output tensor contains internal
+        memory overlaps or if an indexed operation is given duplicate indices.
+        Below are several examples.
+
+    .. code-block:: python
+
+        >>> torch.use_deterministic_algorithms(True)
+
+        # Determinism is not guaranteed for an in-place operation on a tensor
+        # with internal memory overlap
+        >>> a = torch.tensor(1.).as_strided((3,), (0,))
+        >>> a.index_put_((torch.tensor([0, 1, 2]),), torch.tensor([0., 1., 2.]))
```

Comment:
An inplace add instead of index_put_ might demonstrate this point better since add is a simpler operation

Reply:
We should definitely not use `index_put_` since I'm going to add a nondeterministic alert for it.

The trouble is that many operations are now checking for overlaps and throwing an error, so it might be hard to find an example of a simple operation:
```
>>> a = torch.tensor([1.]).as_strided((3,), (0,))
>>> a.add_(torch.tensor([1., 2., 3.]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: unsupported operation: more than one element of the written-to tensor refers to a single memory location. Please clone() the tensor before performing the operation.
```

More operations will likely check this as time goes on. Actually, late last year I was working on a PR to add it to every inplace and out operation in codegen, but I took a break on it and haven't picked it back up yet. Since the goal is to have every out and inplace operation check for overlaps, perhaps we should consider not including an example at all since any currently working example is bound to go out of date at some point?

Reply:
Your analysis makes a lot of sense. Sounds good.